### ISSUE TITLE
React 19 support

### ISSUE THREAD

**Issue by osdiab** [airbnb/visx#1883]
It looks like the peer dependencies of Visx do not include React 19, but [NextJS 15 was released as stable](https://nextjs.org/blog/next-15) and it depends on React 19. Would be great to know if visx is tested against it. Thanks!

**Comment by nevnein** [comment]
I am migrating a project to React 19, right now everything visx-related kinda works already, the only problem I've found is in `useTooltipInPortal` causing an `Uncaught Error: Switched to client rendering because the server rendering errored: window is not defined`. Probably the trick used [here](https://github.com/airbnb/visx/blob/d920798cb9e750477342bbfface99e4399a2d5b8/packages/visx-tooltip/src/hooks/useTooltipInPortal.tsx#L58) to check for SSR is not valid anymore.

**Comment by V-iktor** [comment]
I was getting the following after upgrading

`Cannot find module '@visx/vendor/d3-array' or its corresponding type declarations.ts(2307)`

Never had it in my package.json but adding it manually seems to have solved it and I am not getting any errors

`pnpm i @visx/vendor`

**Comment by darthmaim** [comment]
React 19 is now stable.

The only issue I've encountered so far would be fixed by #1837.

**Comment by williaster** [comment]
Thanks everyone - we weren't sure exactly how to handle this when `Next` upgraded to `React 19`, before it was stable. 

We'll try and get on this now that it's stable, thanks for the investigation and PR @darthmaim 🙏 👀

**Comment by darthmaim** [comment]
I've tried to create a PR to update all peerDependencies to include react@19, but all test fail for various reasons:
- it seems that `enzyme` can't render using react 19 at all (it already doesn't really support react@18 and react@17).
- react@19 uses `queueMicrotask` which is not supported by the 5 year old version of jest used in this project
- upgrading jest fails because of outdated babel
- and so on...

The best way forward is probably to include react@19 in the peer dependencies, but still run all tests in this repo using react@18. Release this is as a prerelease, so it can be "integration tested" by others.

Then slowly work on upgrading all dependencies (which includes rewriting all tests using `@testing-library/react` instead of `enzyme`) to be able to run tests using react@19.

**Comment by williaster** [comment]
Thanks @darthmaim ! 

Will try to get eyes on these asap, agree with your proposed order of events. Will see if I can get the alpha branch out today. Have been discussing expanding some internal LLM scripts to help us migrate the enzyme tests to RTL, may spill into next year but hoping they can do the heavy lifting 💪

**Comment by williaster** [comment]
fyi we got the pre-release out as [3.13.0-alpha.0](https://www.npmjs.com/package/@visx/shape/v/3.13.0-alpha.0), let me know if you see any issues. working through a bunch of other items to get the project itself on `react@19` with great help from @darthmaim !

- [x] https://github.com/airbnb/visx/pull/1889
- [x] https://github.com/airbnb/visx/pull/1890
- [x] https://github.com/airbnb/visx/pull/1891
- [x] https://github.com/airbnb/visx/pull/1893
- [ ] WIP https://github.com/airbnb/visx/pull/1892 / https://github.com/airbnb/visx/pull/1894

**Comment by timrutter** [comment]
has there been any further progress on this issue?

**Comment by BearCooder** [comment]
Hey @williaster, @kristw  & @hshoff sorry for the ping but when can we expect to get a stable release with react 19 support? Its been a while since the the pre-release is out as [3.13.0-alpha.0](https://www.npmjs.com/package/@visx/shape/v/3.13.0-alpha.0) in december.
Thanks!

**Comment by jacobmoshipco** [comment]
I'll add that I've also not been seeing issues using 3.13.2-alpha.0, other than a dependency conflict that was resolved by adding

```json
"overrides": {
    "@visx/xychart": {
        "react-spring": "^10.0.0",
        "@react-spring/web": "^10.0.0"
    }
}
```

**Comment by williaster** [comment]
Sorry all - we are blocked on releasing a stable version by https://github.com/airbnb/visx/pull/1894. Upgrading TS/jest broke our automatic documentation generation and we haven't had bandwidth to debug further unfortunately.

If someone from the community has the expertise and interest, help would be very appreciated on this 🙏

**Comment by pergamonster** [comment]
@williaster I took a look at this, but wasnt able to reproduce the errors mentioned  in #1894  ie 

Check failure on line 28 in scripts/updateTsReferences.ts


GitHub Actions
/ build

'error' is of type 'unknown'.

is that what you were referring to by broke our automatic doc generation above? To investigate i forked the repo and was able to run the pull request action once i updated actions/cache to v4 with all the changes from chrisw--tsc-4.

[chrisw--tsc-4 merged on my fork](https://github.com/pergamonster/visx/actions/runs/15457089368/job/43511218466)

NB Site generation step is success

Could you pls let me know if this helps, many thanks

**Comment by BearCooder** [comment]
Awesome @pergamonster!! Please @williaster, @kristw & @hshoff let pergamonster know if this helps you with releasing the newe version. Thank you

**Comment by snilli** [comment]
why don't include react 19 ? 
[#https://github.com/Taylor-CCB-Group/MDV/pull/244](https://github.com/Taylor-CCB-Group/MDV/pull/244)

**Comment by hshoff** [comment]
Thanks all for reaching out. React 19 is a priority for us. We're on our way there with a [`4.0.0` milestone](https://github.com/airbnb/visx/milestone/97?closed=1).

Last week I unblocked the [TypeScript 5 update](https://github.com/airbnb/visx/pull/1938), updated node, yarn, next, types, migrated to vitest from jest, and generally just modernizing the codebase as it's starting to show it's age 

Next I'll get to the [React 19 deprecations](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#removed-deprecated-react-apis) and dropping React 15, 16 support.

If you encounter React 19 related issues with the 3.*-alpha pre-releases. Please open a new issue with details. Will help me keep track.

We'll get there brick by brick. 

Thanks all for the patience

Keep an eye on [`4.0.0` milestone](https://github.com/airbnb/visx/milestone/97?closed=1)

**Comment by rarenal** [comment]
Tested it in our integration using 4.0.0-alpha.1 both in React 18 & 19 and went through all our checks and everything worked fine.
Packages checked: @visx/axis, @visx/brush, @visx/event, @visx/grid, @visx/group, @visx/hierarchy, @visx/scale, @visx/shape

Thx for the effort @hshoff! looking forward to the release 👍

**Comment by BearCooder** [comment]
Finally had some time to update and test alpha v4 and all works with no issues, awesome work hope to see soon a stable release!

**Comment by artsiompeshko** [comment]
Hi, thanks for the migration!! Any idea on when 4x will become stable? 🙏

**Comment by ijsje71** [comment]
Is there any timeline as to when the 4.x.x releases will become stable?

**Comment by hshoff** [comment]
visx 4 is out. start here: https://github.com/airbnb/visx/blob/master/MIGRATION.md
